extensions/data-workspace/src/services/workspaceService.ts

```diff
@@ -265,20 +266,44 @@ export class WorkspaceService implements IWorkspaceService {
 			if (projectTypes && projectTypes.length > 0) {
 				if (projType) {
 					if (projectTypes.findIndex((proj: string) => proj.toUpperCase() === projType) !== -1) {
-						await this.handleProjectProviderExtension(extension);
+						await this.handleProjectProviderExtension(extension, forceActivate);
 						break;
 					}
 				} else {
-					await this.handleProjectProviderExtension(extension);
+					await this.handleProjectProviderExtension(extension, forceActivate);
 				}
 			}
 		}
 	}
 
-	private async handleProjectProviderExtension(extension: vscode.Extension<any>): Promise<void> {
+	/**
+	 * Ensures project provider extension is activated and registered
+	 * @param extension Extension to activate and register
+	 * @param forceActivate Activates the extension if true. If false, only activates the extension if the workspace contains a file with the extension's project type
+	 */
+	private async handleProjectProviderExtension(extension: vscode.Extension<any>, forceActivate: boolean = false): Promise<void> {
 		try {
 			if (!extension.isActive) {
-				await extension.activate();
+				if (forceActivate) {
+					await extension.activate();
+				} else {
+					const projectTypes = extension.packageJSON.contributes?.projects as string[] | undefined;
+					if (!projectTypes) {
+						return;
+					}
+
+					let projFilesInWorkspace: vscode.Uri[] = [];
+					for (const projType of projectTypes) {
+						(await vscode.workspace.findFiles(`**/*.${projType}`)).forEach(f => projFilesInWorkspace.push(f));
+					}
+
+					if (projFilesInWorkspace?.length > 0) {
+						// only try to activate the extension if the workspace has a project with that extension
+						await extension.activate();
+					} else {
+						return;
+					}
```

Comment:
This can be simplified - you only need to find one project that matches to justify activating the extension and then can skip checking all the other supported types at that point. 

```suggestion
					for (const projType of projectTypes) {
						const projFilesInWorkspace = await vscode.workspace.findFiles(`**/*.${projType}`);
						if (projFilesInWorkspace.length > 0) {
							// only try to activate the extension if the workspace has at least one project with that extension
							await extension.activate();
							break;
						}
					}
```